sdcm/sct_events/event_counter.py

```diff
@@ -0,0 +1,192 @@
+
+    def run(self) -> None:
+        LOGGER.debug("Counting events is running")
+        for event_tuple in self.inbound_events():
+            if not self._start_count.is_set():
+                continue
+
+            with verbose_suppress("EventsFileLogger failed to process %s", event_tuple):
+                _, event = event_tuple  # try to unpack event from EventsDevice
+
+                for cm_id, event_stat_data in self._register.items():
+                    if event_stat_data.event_types and event.__class__ in event_stat_data.event_types:
+                        handler = EventStatHandleMapper.get(event.__class__.__name__, EventStatHandler)
+                        stat = event_stat_data.stats.setdefault(event.__class__.__name__, {})
+                        event_stat_data.stats.update({event.__class__.__name__: handler(
+                            event, event_stat_data.save_dir).update_stat(stat)})
+                        self._register[cm_id] = event_stat_data
+
+    def start_counter(self):
+        if not self._start_count.is_set():
+            self._start_count.set()
+
+    def stop_counter(self):
+        if not self._register:
+            self._start_count.clear()
+
+    def add_counter(self, counter_id: str, event_stat_data: EventStatData):
+        self._register[counter_id] = event_stat_data
+
+    def get_counter(self, counter_id: str) -> EventStatData | None:
+        if counter_id in self._register:
+            return self._register[counter_id]
+        return None
+
+    def remove_counter(self, counter_id: str):
+        if counter_id in self._register:
+            del self._register[counter_id]
+
+
+start_events_counter = partial(start_events_process, EVENTS_COUNTER_ID, EventsCounter)
+get_events_counter = cast(Callable[..., EventsCounter], partial(get_events_process, EVENTS_COUNTER_ID))
+
+
+class EventCounterContextManager:
+    def __init__(self, event_type, name, _registry=None):
+        self._events_stat_dir = get_events_counter(_registry=_registry).events_stat_dir
+        self._id = f"{uuid4()}"
+
+        if not isinstance(event_type, tuple):
+            event_type = (event_type, )
+
+        self._event_type = event_type
+        self._counter_device = get_events_counter(_registry=_registry)
+        self._statistics = {}
+        self._save_file_path = self._events_stat_dir / Path(name)
+
+    def start_event_counter(self):
+        event_stat_data = EventStatData(event_types=self._event_type,
+                                        save_dir=self._save_file_path,
+                                        stats={})
+        self._counter_device.add_counter(self._id, event_stat_data)
+        self._counter_device.start_counter()
+
+    def stop_event_counter(self):
+        if counter_data := self._counter_device.get_counter(self._id):
+            self._statistics = counter_data.stats
+            self._counter_device.remove_counter(self._id)
+        self._counter_device.stop_counter()
```

Comment:
What's the point of having different counters if on `__exit__` we make  `EventsCounter` not to count?
Why do we need different counters? Maybe it would be enough to get initial count value on enter and get diff on exit instead adding different counters?

Reply:
I think start/stop counter is not only redundant in that case, but also may lead to errors when within `EventCounterContextManager` someone will open another `EventCounterContextManager` with different counter - then on exit will stop counting for all counters.

I think we could just drop `self._start_count` Event idea and count only if register contain counters, otherwise, skip any work.

Reply:
> I think start/stop counter is not only redundant in that case, but also may lead to errors when within `EventCounterContextManager` someone will open another `EventCounterContextManager` with different counter - then on exit will stop counting for all counters.
> 

stop counting will happened , only if no any registered context managers will stay in _register. So if some on open and close counterevent_cm, it will not affect on others, and only latest closed counter_cm will stop counting

> I think we could just drop `self._start_count` Event idea and count only if register contain counters, otherwise, skip any work.

It is interesting idea, will check it now

Reply:
@soyacz , removed the _event. Added 2 more unit tests could take a look